lib/pos_blockchain.go

```diff
@@ -130,12 +144,97 @@ func (bc *Blockchain) validateDeSoBlockPoS(desoBlock *MsgDeSoBlock) error {
-func (bc *Blockchain) validateBlockGeneral(desoBlock *MsgDeSoBlock) error {
-	// TODO: Implement me
-	return errors.New("IMPLEMENT ME")
+func (bc *Blockchain) validateBlockIntegrity(desoBlock *MsgDeSoBlock) error {
+	// First make sure we have a non-nil header
+	if desoBlock.Header == nil {
+		return RuleErrorNilBlockHeader
+	}
+
+	// Make sure we have a prevBlockHash
+	if desoBlock.Header.PrevBlockHash == nil {
+		return RuleErrorNilPrevBlockHash
+	}
+
+	// Timestamp validation
+
+	// Validate that the timestamp is not less than its parent.
+	parentBlock, exists := bc.blockIndex[*desoBlock.Header.PrevBlockHash]
+	if !exists {
+		// Note: this should never happen as we only call this function after
+		// we've validated that all ancestors exist in the block index.
+		return RuleErrorMissingParentBlock
+	}
+	if desoBlock.Header.TstampNanoSecs < parentBlock.Header.TstampNanoSecs {
+		return RuleErrorPoSBlockTstampNanoSecsTooOld
+	}
+	// TODO: Add support for putting the drift into global params.
+	if desoBlock.Header.TstampNanoSecs > uint64(time.Now().UnixNano())+bc.params.DefaultBlockTimestampDriftNanoSecs {
+		return RuleErrorPoSBlockTstampNanoSecsInFuture
+	}
+
+	// Header validation
+	if desoBlock.Header.Version != HeaderVersion2 {
+		return RuleErrorInvalidPoSBlockHeaderVersion
+	}
+
+	// Malformed block checks
+	// Require header to have either vote or timeout QC
+	isTimeoutQCEmpty := desoBlock.Header.ValidatorsTimeoutAggregateQC.isEmpty()
+	isVoteQCEmpty := desoBlock.Header.ValidatorsVoteQC.isEmpty()
+	if isTimeoutQCEmpty && isVoteQCEmpty {
+		return RuleErrorNoTimeoutOrVoteQC
+	}
+
+	if !isTimeoutQCEmpty && !isVoteQCEmpty {
+		return RuleErrorBothTimeoutAndVoteQC
+	}
+
+	if !isVoteQCEmpty && len(desoBlock.Txns) == 0 {
+		return RuleErrorVoteQCWithoutTransactions
+	}
+
+	if !isTimeoutQCEmpty && len(desoBlock.Txns) != 0 {
+		return RuleErrorTimeoutQCWithTransactions
+	}
+
+	if desoBlock.Header.ProposerVotingPublicKey.IsEmpty() {
+		return RuleErrorInvalidProposerVotingPublicKey
+	}
+
+	if desoBlock.Header.ProposerPublicKey == nil || desoBlock.Header.ProposerPublicKey.IsZeroPublicKey() {
+		return RuleErrorInvalidProposerPublicKey
+	}
+
+	if desoBlock.Header.ProposerRandomSeedHash.isEmpty() {
+		return RuleErrorInvalidRandomSeedHash
+	}
+
+	merkleRoot := desoBlock.Header.TransactionMerkleRoot
+
+	// We only want to check the merkle root if we have more than 0 transactions.
+	if isTimeoutQCEmpty {
+		if merkleRoot == nil {
+			return RuleErrorNilMerkleRoot
+		}
+		computedMerkleRoot, _, err := ComputeMerkleRoot(desoBlock.Txns)
+		if err != nil {
+			return errors.Wrapf(err, "validateBlockIntegrity: Problem computing merkle root")
+		}
+		if !merkleRoot.IsEqual(computedMerkleRoot) {
+			return RuleErrorInvalidMerkleRoot
+		}
+	} else {
+		// TODO: do we want to ensure a nil merkle root if we have a timeout QC?
+		if merkleRoot != nil {
+			return RuleErrorTimeoutQCWithMerkleRoot
+		}
+	}
```

Comment:
Should this explicitly check if the block has one or more txns? See my comment above on why it may be fine to have a block with 0 txns.